compiler/rustc_hir_analysis/src/check/method/suggest.rs

```diff
@@ -1201,6 +1204,81 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         false
     }
 
+    fn suggest_range_for_iter(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        actual: Ty<'tcx>,
+        source: SelfSource<'tcx>,
+        span: Span,
+        item_name: Ident,
+        ty_str: &str,
+    ) -> bool {
+        if let SelfSource::MethodCall(expr) = source {
+            for (_, parent) in tcx.hir().parent_iter(expr.hir_id).take(5) {
+                if let Node::Expr(parent_expr) = parent {
+                    let lang_item = match parent_expr.kind {
+                        ExprKind::Struct(ref qpath, _, _) => match **qpath {
+                            QPath::LangItem(LangItem::Range, ..) => Some(LangItem::Range),
+                            QPath::LangItem(LangItem::RangeTo, ..) => Some(LangItem::RangeTo),
+                            QPath::LangItem(LangItem::RangeToInclusive, ..) => {
+                                Some(LangItem::RangeToInclusive)
+                            }
+                            _ => None,
+                        },
+                        ExprKind::Call(ref func, _) => match func.kind {
+                            // `..=` desugars into `::std::ops::RangeInclusive::new(...)`.
+                            ExprKind::Path(QPath::LangItem(LangItem::RangeInclusiveNew, ..)) => {
+                                Some(LangItem::RangeInclusiveStruct)
+                            }
+                            _ => None,
+                        },
+                        _ => None,
+                    };
+
+                    if lang_item.is_none() {
+                        continue;
+                    }
+
+                    let span_included = match parent_expr.kind {
+                        hir::ExprKind::Struct(_, eps, _) => {
+                            eps.len() > 0 && eps.last().map_or(false, |ep| ep.span.contains(span))
+                        }
+                        // `..=` desugars into `::std::ops::RangeInclusive::new(...)`.
+                        hir::ExprKind::Call(ref func, ..) => func.span.contains(span),
+                        _ => false,
+                    };
+
+                    if !span_included {
+                        continue;
+                    }
+
+                    debug!("lang_item: {:?}", lang_item);
+                    let range_def_id = self.tcx.require_lang_item(lang_item.unwrap(), None);
+                    let range_ty =
+                        self.tcx.bound_type_of(range_def_id).subst(self.tcx, &[actual.into()]);
+
+                    let pick =
+                        self.lookup_probe(span, item_name, range_ty, expr, ProbeScope::AllTraits);
```

Comment:
can you emit a comment for why we're using `ProbeScope::AllTraits` here, I remember that you mentioned the reason for that before but I forgot that again :grin: 

also, i think ideally we would use `probe_for_name` directly here with `IsSuggestion(true)`. Not sure what `IsSuggestion` does but we are emitting a suggestion here, so setting that to true seems like it should be done :sweat_smile:

Reply:
Seems `IsSuggestion` only used at here:
https://github.com/rust-lang/rust/blob/1536ab1b383f21b38f8d49230a2aecc51daffa3d/compiler/rustc_hir_analysis/src/check/method/probe.rs#L1230

Use `ProbeScope::TraitsInScope` can not find the pick, but `ProbeScope::AllTraits` will find it.

Reply:
it seems helpful to not emit `unstable_name_collision_hint`s while looking for diagnostics, so that should still be done.

`ProbeScope::TraitsInScope` that feels weird and shouldn't be the case as `Iterator` is in scope. Is this still the case with the current code?

Reply:
updated the code to use `probe_for_name`.

`ProbeScope::TraitsInScope` means find with the scope of `expr`, 
https://github.com/rust-lang/rust/blob/1536ab1b383f21b38f8d49230a2aecc51daffa3d/compiler/rustc_hir_analysis/src/check/method/probe.rs#L456

othewise try to find methods in all traits.

At this function, we are at a point of error handling of range type checking, since error happened so the `Iterator` is not in the scope? it's only my guess😁